CRM/Financial/BAO/FinancialItem.php

```diff
@@ -286,25 +286,23 @@ public static function checkContactPresent($contactIds, &$error) {
   }
 
   /**
-   * Get last financial item data.
+   * Get non-liability financial item with the highest id relating to the line item.
    *
-   * @param int $entityId
+   * In practical terms this means return the row that is not associated with
+   * sales tax as sales tax is a liability account.
    *
-   * @param string $entityTable
+   * @param int $entityId
    *
    * @return object CRM_Core_DAO
    */
-  public static function getPreviousFinancialItem($entityId, $entityTable = 'civicrm_line_item') {
-    $queryParams = array(
-      1 => array($entityId, 'Integer'),
-      2 => array($entityTable, 'String'),
-    );
-    $query = 'SELECT id, description, status_id, financial_account_id 
-      FROM civicrm_financial_item
-      WHERE entity_id = %1 AND entity_table = %2 ORDER BY id DESC LIMIT 1';
-    $prevFinancialItem = CRM_Core_DAO::executeQuery($query, $queryParams);
-    $prevFinancialItem->fetch();
-    return $prevFinancialItem;
+  public static function getPreviousNonLiabilityFinancialItem($entityId) {
+    $liabilityFinancialAccounts = civicrm_api3('FinancialAccount', 'get', array('financial_account_type_id' => 'Liability'));
```

Comment:
I looked at this sql & the proposed sql & it seemed to me to boil down to this (which does make the tests pass)